antarest/storage/service.py

```diff
@@ -38,52 +41,62 @@ def __init__(
+        uuid, url = StorageServiceUtils.extract_info_from_url(route)
+        md = self._get_metadata_according_permission(params.user, uuid)
 
-        return self.study_service.get(route, depth)
+        return self.study_service.get(md, url, depth)
 
-    def _get_study_uuids(self, params: RequestParameters) -> List[str]:
-        uuids = self.study_service.get_study_uuids()
-        return [
-            uuid
-            for uuid in uuids
-            if self._check_user_permission(params.user, uuid, raising=False)
-        ]
+    def _get_study_metadatas(
+        self, params: RequestParameters
+    ) -> List[Metadata]:
+        metadatas: List[Metadata] = list()
+
+        for uuid in self.study_service.get_study_uuids():
+            md = self._get_metadata_according_permission_no_raising(
+                params.user, uuid
+            )
+            if md:
+                metadatas.append(md)
+        return metadatas
 
     def get_studies_information(self, params: RequestParameters) -> JSON:
-        uuids = self._get_study_uuids(params)
         return {
-            uuid: self.study_service.get_study_information(uuid)
-            for uuid in uuids
+            md.id: self.study_service.get_study_information(md)
+            for md in self._get_study_metadatas(params)
         }
 
     def get_study_information(
         self, uuid: str, params: RequestParameters
     ) -> JSON:
-        self._check_user_permission(params.user, uuid)
-        return self.study_service.get_study_information(uuid)
+        md = self._get_metadata_according_permission(params.user, uuid)
+        return self.study_service.get_study_information(md)
 
     def get_study_path(self, uuid: str, params: RequestParameters) -> Path:
-        self._check_user_permission(params.user, uuid)
-        return self.study_service.get_study_path(uuid)
+        md = self._get_metadata_according_permission(params.user, uuid)
+        return self.study_service.get_study_path(md)
 
     def create_study(self, study_name: str, params: RequestParameters) -> str:
-        uuid = self.study_service.create_study(study_name)
-        self._save_metadata(uuid, params.user)
-        return uuid
+        md = Metadata(id=str(uuid4()), name=study_name, workspace="default")
+        md = self.study_service.create_study(md)
+        self._save_metadata(md, params.user)
+        return str(md.id)
 
     def copy_study(
         self,
         src_uuid: str,
         dest_study_name: str,
         params: RequestParameters,
     ) -> str:
-        self._check_user_permission(params.user, src_uuid)
-        uuid = self.study_service.copy_study(src_uuid, dest_study_name)
-        self._save_metadata(uuid, params.user)
+        src_md = self._get_metadata_according_permission(params.user, src_uuid)
+
+        dest_md = deepcopy(src_md)
+        dest_md.id = str(uuid4())
+        dest_md.name = dest_study_name
+
+        md = self.study_service.copy_study(src_md, dest_md)
```

Comment:
why the creation of the metadata for the new study not done in copy_study (with second arg being the new name) ?

Reply:
also copy_study should have access to metadata repository so it will also save the metadata which is an integral part of the study creation/copy process

Reply:
Metadata object is managed by StorageService, Begin to create metadata elsewhere in code (and generate its id outside StorageService) create a less clear separate architecture.